### ISSUE TITLE
Incorrect documentation for role based access

### ISSUE THREAD

**Issue by ralphsmith80** [firebase/snippets-rules#4]
I'm creating this after getting confirmation from support that it is a reproducible issue.

## Problem
After implementing this myself I believe the [role based access documents](https://firebase.google.com/docs/firestore/solutions/role-based-access) is incorrect. 

On [line 28 in step 3](https://github.com/firebase/snippets-rules/blob/master/rules/solution-rbac/step3.rules#L28) the rule will invoke the following check `resource.data.roles[request.auth.uid]` for a comment. The issue that I've noted is that the `resource` which will be a `comment` at this point does not have a `roles` map. Which causes a failure since there is no query that can be written to align with this rule. 

What I've done to implement this is add a `getRoles` function and modify the parameters of `getRole` and `isOneOfRoles` to encapsulate writing rules for both stories and comments. Here's the a snippet of what I've used. 

``` 
service cloud.firestore {
  match /databases/{database}/documents {
    match /stories/{story} {
      function isSignedIn() {
        return request.auth != null;
      }
      function getRoles() {
        // Get the "roles" map in the story document.
        return get(/databases/$(database)/documents/stories/$(story)).data.roles;
      }
      function getRole(roles) {
        // Read from the "roles" map in the story document.
        return roles[request.auth.uid];
      }
      function isOneOfRoles(role, array) {
        // Determine if the user is one of an array of roles
        return isSignedIn() && (role in array);
      }
      function isValidNewStory() {
        return resource.data == null
          && request.resource.data.roles[request.auth.uid] == 'owner';
      }

      allow write: if isValidNewStory() || isOneOfRoles(getRole(resource.data.roles), ['owner']);

      // Any role can read stories.
      allow read: if isOneOfRoles(getRole(resource.data.roles), ['owner', 'writer', 'commenter', 'reader']);

      match /comments/{comment} {
        // Any role can read comments.
        allow read: if isOneOfRoles(getRole(getRoles()), ['owner', 'writer', 'commenter', 'reader']);
      }
    }
  }
}
``` 
## Suggested fix from support - (doesn't work)

read - **fails**
write - **works**

```
function getRole() {
  //return resource.data.roles[request.auth.uid];
  return get(/databases/$(database)/documents/stories/$(story)).data.roles[request.auth.uid];
}
```

I think this was my original approach. Unfortunately it doesn't work and I can not explain why. Logically they should be equivalent but I think the rules are getting interpreted differently somewhere in the Firestore pipeline. 

Having said that I've taken note of this [Stackoverflow issue][so-issue] while working through some other related issues. While the issue might provide some insight, similar workaround rules does work.

## Alternative solutions that still don't work

read - **fails**
write - **works**

```
function getRole() {
  return get(/databases/$(database)/documents/stories/$(story)).data.roles[request.auth.uid]
    || resource.data.roles[request.auth.uid];
}
```

Based on this [Stackoverflow issue][so-issue]
read - **fails**
write - **works**

```
function getRole() {
  return (
    resource.data.roles[request.auth.uid]
    || resource.roles[request.auth.uid]
    || get(/databases/$(database)/documents/stories/$(story)).data.roles[request.auth.uid]
    || get(/databases/$(database)/documents/stories/$(story)).roles[request.auth.uid]
  );
}
```

### Realtime Listener
I'm using the realtime listener to get updates and I've applied a `where` clause to align with the read rule.

```
db
  .collection('stories')
  .where(`roles.${user.uid}`, '==', 'owner')
  .onSnapshot(querySnapshot => {
    querySnapshot.forEach(doc => {
      if (!doc.exists) return
      console.log(doc.data())
    },
    error => console.error(error),
  )
```

<details>
  <summary>**Error**</summary>
Error: Missing or insufficient permissions.
    at new FirestoreError (error.js:149)
    at JsonProtoSerializer../node_modules/@firebase/firestore/dist/esm/src/remote/serializer.js.JsonProtoSerializer.fromRpcStatus (serializer.js:93)
    at JsonProtoSerializer../node_modules/@firebase/firestore/dist/esm/src/remote/serializer.js.JsonProtoSerializer.fromWatchChange (serializer.js:573)
    at PersistentListenStream../node_modules/@firebase/firestore/dist/esm/src/remote/persistent_stream.js.PersistentListenStream.onMessage (persistent_stream.js:391)
    at persistent_stream.js:333
    at persistent_stream.js:308
    at async_queue.js:83
    at <anonymous>
__stack_frame_overlay_proxy_console__ @ index.js:2177
(anonymous) @ StoriesListener.js:33
(anonymous) @ async_observer.js:44
setTimeout (async)
./node_modules/@firebase/firestore/dist/esm/src/util/async_observer.js.AsyncObserver.scheduleEvent @ async_observer.js:42
./node_modules/@firebase/firestore/dist/esm/src/util/async_observer.js.AsyncObserver.error @ async_observer.js:34
./node_modules/@firebase/firestore/dist/esm/src/core/event_manager.js.QueryListener.onError @ event_manager.js:167
./node_modules/@firebase/firestore/dist/esm/src/core/event_manager.js.EventManager.onError @ event_manager.js:106
(anonymous) @ sync_engine.js:318
Promise.then (async)
(anonymous) @ sync_engine.js:317
Promise.then (async)
./node_modules/@firebase/firestore/dist/esm/src/core/sync_engine.js.SyncEngine.rejectListen @ sync_engine.js:316
(anonymous) @ remote_store.js:453
Promise.then (async)
(anonymous) @ remote_store.js:450
./node_modules/@firebase/firestore/dist/esm/src/remote/remote_store.js.RemoteStore.handleTargetError @ remote_store.js:449
./node_modules/@firebase/firestore/dist/esm/src/remote/remote_store.js.RemoteStore.onWatchStreamChange @ remote_store.js:328
./node_modules/@firebase/firestore/dist/esm/src/remote/persistent_stream.js.PersistentListenStream.onMessage @ persistent_stream.js:393
(anonymous) @ persistent_stream.js:333
(anonymous) @ persistent_stream.js:308
(anonymous) @ async_queue.js:83
Promise.then (async)
./node_modules/@firebase/firestore/dist/esm/src/util/async_queue.js.AsyncQueue.scheduleInternal @ async_queue.js:81
./node_modules/@firebase/firestore/dist/esm/src/util/async_queue.js.AsyncQueue.schedule @ async_queue.js:76
dispatchIfStillActive @ persistent_stream.js:305
(anonymous) @ persistent_stream.js:332
./node_modules/@firebase/firestore/dist/esm/src/remote/stream_bridge.js.StreamBridge.callOnMessage @ stream_bridge.js:55
(anonymous) @ webchannel_connection.js:255
(anonymous) @ webchannel_connection.js:195
Ab @ index.js:22
./node_modules/@firebase/webchannel-wrapper/dist/index.js.g.dispatchEvent @ index.js:20
./node_modules/@firebase/webchannel-wrapper/dist/index.js.Pe.Ca @ index.js:95
./node_modules/@firebase/webchannel-wrapper/dist/index.js.g.Ma @ index.js:82
bd @ index.js:41
$c @ index.js:38
Yc @ index.js:37
./node_modules/@firebase/webchannel-wrapper/dist/index.js.g.Qa @ index.js:36
./node_modules/@firebase/webchannel-wrapper/dist/index.js.g.hb @ index.js:35
Ab @ index.js:22
./node_modules/@firebase/webchannel-wrapper/dist/index.js.g.dispatchEvent @ index.js:20
te @ index.js:66
./node_modules/@firebase/webchannel-wrapper/dist/index.js.g.cb @ index.js:65
./node_modules/@firebase/webchannel-wrapper/dist/index.js.g.La @ index.js:65
XMLHttpRequest.send (async)
./node_modules/@firebase/webchannel-wrapper/dist/index.js.g.fa @ index.js:63
Uc @ index.js:35
Vc @ index.js:34
./node_modules/@firebase/webchannel-wrapper/dist/index.js.g.Na @ index.js:78
Vb @ index.js:28
Promise.then (async)
Tb @ index.js:28
Ie @ index.js:76
./node_modules/@firebase/webchannel-wrapper/dist/index.js.g.Ma @ index.js:82
bd @ index.js:41
$c @ index.js:38
Yc @ index.js:37
./node_modules/@firebase/webchannel-wrapper/dist/index.js.g.Qa @ index.js:36
./node_modules/@firebase/webchannel-wrapper/dist/index.js.g.hb @ index.js:35
Ab @ index.js:22
./node_modules/@firebase/webchannel-wrapper/dist/index.js.g.dispatchEvent @ index.js:20
te @ index.js:66
./node_modules/@firebase/webchannel-wrapper/dist/index.js.g.cb @ index.js:65
./node_modules/@firebase/webchannel-wrapper/dist/index.js.g.La @ index.js:65
XMLHttpRequest.send (async)
./node_modules/@firebase/webchannel-wrapper/dist/index.js.g.fa @ index.js:63
Uc @ index.js:35
Sc @ index.js:34
./node_modules/@firebase/webchannel-wrapper/dist/index.js.g.Oa @ index.js:73
(anonymous) @ index.js:33
setTimeout (async)
Ec @ index.js:33
De @ index.js:71
Sd @ index.js:79
Rd @ index.js:54
./node_modules/@firebase/webchannel-wrapper/dist/index.js.g.Za @ index.js:93
sendFn @ webchannel_connection.js:174
./node_modules/@firebase/firestore/dist/esm/src/remote/stream_bridge.js.StreamBridge.send @ stream_bridge.js:43
./node_modules/@firebase/firestore/dist/esm/src/remote/persistent_stream.js.PersistentStream.sendRequest @ persistent_stream.js:204
./node_modules/@firebase/firestore/dist/esm/src/remote/persistent_stream.js.PersistentListenStream.watch @ persistent_stream.js:409
./node_modules/@firebase/firestore/dist/esm/src/remote/remote_store.js.RemoteStore.sendWatchRequest @ remote_store.js:250
(anonymous) @ remote_store.js:300
forEachNumber @ obj.js:41
./node_modules/@firebase/firestore/dist/esm/src/remote/remote_store.js.RemoteStore.onWatchStreamOpen @ remote_store.js:299
(anonymous) @ persistent_stream.js:323
(anonymous) @ persistent_stream.js:308
(anonymous) @ async_queue.js:83
Promise.then (async)
./node_modules/@firebase/firestore/dist/esm/src/util/async_queue.js.AsyncQueue.scheduleInternal @ async_queue.js:81
./node_modules/@firebase/firestore/dist/esm/src/util/async_queue.js.AsyncQueue.schedule @ async_queue.js:76
dispatchIfStillActive @ persistent_stream.js:305
(anonymous) @ persistent_stream.js:320
./node_modules/@firebase/firestore/dist/esm/src/remote/stream_bridge.js.StreamBridge.callOnOpen @ stream_bridge.js:47
(anonymous) @ webchannel_connection.js:264
setTimeout (async)
./node_modules/@firebase/firestore/dist/esm/src/platform_browser/webchannel_connection.js.WebChannelConnection.openStream @ webchannel_connection.js:259
./node_modules/@firebase/firestore/dist/esm/src/remote/persistent_stream.js.PersistentListenStream.startRpc @ persistent_stream.js:386
./node_modules/@firebase/firestore/dist/esm/src/remote/persistent_stream.js.PersistentStream.startStream @ persistent_stream.js:317
(anonymous) @ persistent_stream.js:280
e.g @ auth.js:23
Yb @ auth.js:26
Ub @ auth.js:26
./node_modules/@firebase/auth/dist/auth.js.h.Qb @ auth.js:25
Cb @ auth.js:19
Promise.then (async)
zb @ auth.js:18
yb @ auth.js:18
Vb @ auth.js:25
Rb @ auth.js:22
Pb @ auth.js:23
./node_modules/@firebase/auth/dist/auth.js.z.then @ auth.js:21
./node_modules/@firebase/auth/dist/auth.js.h.Vb @ auth.js:257
./node_modules/@firebase/firestore/dist/esm/src/api/credentials.js.FirebaseCredentialsProvider.getToken @ credentials.js:93
./node_modules/@firebase/firestore/dist/esm/src/remote/persistent_stream.js.PersistentStream.auth @ persistent_stream.js:276
./node_modules/@firebase/firestore/dist/esm/src/remote/persistent_stream.js.PersistentStream.start @ persistent_stream.js:156
./node_modules/@firebase/firestore/dist/esm/src/remote/remote_store.js.RemoteStore.startWatchStream @ remote_store.js:272
./node_modules/@firebase/firestore/dist/esm/src/remote/remote_store.js.RemoteStore.listen @ remote_store.js:226
(anonymous) @ sync_engine.js:136
Promise.then (async)
(anonymous) @ sync_engine.js:126
Promise.then (async)
(anonymous) @ sync_engine.js:123
Promise.then (async)
./node_modules/@firebase/firestore/dist/esm/src/core/sync_engine.js.SyncEngine.listen @ sync_engine.js:120
./node_modules/@firebase/firestore/dist/esm/src/core/event_manager.js.EventManager.listen @ event_manager.js:59
(anonymous) @ firestore_client.js:267
(anonymous) @ async_queue.js:83
Promise.then (async)
./node_modules/@firebase/firestore/dist/esm/src/util/async_queue.js.AsyncQueue.scheduleInternal @ async_queue.js:81
./node_modules/@firebase/firestore/dist/esm/src/util/async_queue.js.AsyncQueue.schedule @ async_queue.js:76
./node_modules/@firebase/firestore/dist/esm/src/core/firestore_client.js.FirestoreClient.listen @ firestore_client.js:266
./node_modules/@firebase/firestore/dist/esm/src/api/database.js.Query.onSnapshotInternal @ database.js:1012
./node_modules/@firebase/firestore/dist/esm/src/api/database.js.Query.onSnapshot @ database.js:993
Listener._this.getRealtime @ StoriesListener.js:11
componentDidMount @ StoriesListener.js:42
commitLifeCycles @ react-dom.development.js:8770
commitAllLifeCycles @ react-dom.development.js:9946
callCallback @ react-dom.development.js:542
invokeGuardedCallbackDev @ react-dom.development.js:581
invokeGuardedCallback @ react-dom.development.js:438
commitRoot @ react-dom.development.js:10050
performWorkOnRoot @ react-dom.development.js:11017
performWork @ react-dom.development.js:10967
requestWork @ react-dom.development.js:10878
scheduleWorkImpl @ react-dom.development.js:10732
scheduleWork @ react-dom.development.js:10689
enqueueSetState @ react-dom.development.js:6212
./node_modules/react/cjs/react.development.js.Component.setState @ react.development.js:237
(anonymous) @ Bootstrap.js:18
Promise.then (async)
componentDidMount @ Bootstrap.js:15
commitLifeCycles @ react-dom.development.js:8770
commitAllLifeCycles @ react-dom.development.js:9946
callCallback @ react-dom.development.js:542
invokeGuardedCallbackDev @ react-dom.development.js:581
invokeGuardedCallback @ react-dom.development.js:438
commitRoot @ react-dom.development.js:10050
performWorkOnRoot @ react-dom.development.js:11017
performWork @ react-dom.development.js:10967
requestWork @ react-dom.development.js:10878
scheduleWorkImpl @ react-dom.development.js:10732
scheduleWork @ react-dom.development.js:10689
scheduleTopLevelUpdate @ react-dom.development.js:11193
updateContainer @ react-dom.development.js:11231
(anonymous) @ react-dom.development.js:15226
unbatchedUpdates @ react-dom.development.js:11102
renderSubtreeIntoContainer @ react-dom.development.js:15225
render @ react-dom.development.js:15290
./src/index.js @ index.js:12
__webpack_require__ @ bootstrap 35471e89bac2241db42c:678
fn @ bootstrap 35471e89bac2241db42c:88
0 @ theme.js:5
__webpack_require__ @ bootstrap 35471e89bac2241db42c:678
./node_modules/@firebase/app/dist/esm/index.js.Object.defineProperty.value @ bootstrap 35471e89bac2241db42c:724
(anonymous) @ bootstrap 35471e89bac2241db42c:724
</details>

### Basic data get

I have attempted this without the realtime listener just to completeness.

```
db
  .collection('stories')
  .where(`roles.${user.uid}`, '==', 'owner')
  .get()
  .then(function(querySnapshot) {
    querySnapshot.forEach(function(doc) {
      // doc.data() is never undefined for query doc snapshots
      console.log(doc.id, ' => ', doc.data())
    })
  })
```

<details>
  <summary>**Error**</summary>
Uncaught (in promise) Error: Missing or insufficient permissions.
    at new FirestoreError (error.js:149)
    at JsonProtoSerializer../node_modules/@firebase/firestore/dist/esm/src/remote/serializer.js.JsonProtoSerializer.fromRpcStatus (serializer.js:93)
    at JsonProtoSerializer../node_modules/@firebase/firestore/dist/esm/src/remote/serializer.js.JsonProtoSerializer.fromWatchChange (serializer.js:573)
    at PersistentListenStream../node_modules/@firebase/firestore/dist/esm/src/remote/persistent_stream.js.PersistentListenStream.onMessage (persistent_stream.js:391)
    at persistent_stream.js:333
    at persistent_stream.js:308
    at async_queue.js:83
    at <anonymous>
FirestoreError @ error.js:149
./node_modules/@firebase/firestore/dist/esm/src/remote/serializer.js.JsonProtoSerializer.fromRpcStatus @ serializer.js:93
./node_modules/@firebase/firestore/dist/esm/src/remote/serializer.js.JsonProtoSerializer.fromWatchChange @ serializer.js:573
./node_modules/@firebase/firestore/dist/esm/src/remote/persistent_stream.js.PersistentListenStream.onMessage @ persistent_stream.js:391
(anonymous) @ persistent_stream.js:333
(anonymous) @ persistent_stream.js:308
(anonymous) @ async_queue.js:83
Promise.then (async)
componentDidMount @ StoriesListener.js:43
commitLifeCycles @ react-dom.development.js:8770
commitAllLifeCycles @ react-dom.development.js:9946
callCallback @ react-dom.development.js:542
invokeGuardedCallbackDev @ react-dom.development.js:581
invokeGuardedCallback @ react-dom.development.js:438
commitRoot @ react-dom.development.js:10050
performWorkOnRoot @ react-dom.development.js:11017
performWork @ react-dom.development.js:10967
requestWork @ react-dom.development.js:10878
scheduleWorkImpl @ react-dom.development.js:10732
scheduleWork @ react-dom.development.js:10689
enqueueSetState @ react-dom.development.js:6212
./node_modules/react/cjs/react.development.js.Component.setState @ react.development.js:237
(anonymous) @ Bootstrap.js:18
Promise.then (async)
componentDidMount @ Bootstrap.js:15
commitLifeCycles @ react-dom.development.js:8770
commitAllLifeCycles @ react-dom.development.js:9946
callCallback @ react-dom.development.js:542
invokeGuardedCallbackDev @ react-dom.development.js:581
invokeGuardedCallback @ react-dom.development.js:438
commitRoot @ react-dom.development.js:10050
performWorkOnRoot @ react-dom.development.js:11017
performWork @ react-dom.development.js:10967
requestWork @ react-dom.development.js:10878
scheduleWorkImpl @ react-dom.development.js:10732
scheduleWork @ react-dom.development.js:10689
scheduleTopLevelUpdate @ react-dom.development.js:11193
updateContainer @ react-dom.development.js:11231
(anonymous) @ react-dom.development.js:15226
unbatchedUpdates @ react-dom.development.js:11102
renderSubtreeIntoContainer @ react-dom.development.js:15225
render @ react-dom.development.js:15290
./src/index.js @ index.js:12
__webpack_require__ @ bootstrap 36e7a445048220614319:678
fn @ bootstrap 36e7a445048220614319:88
0 @ theme.js:5
__webpack_require__ @ bootstrap 36e7a445048220614319:678
./node_modules/@firebase/app/dist/esm/index.js.Object.defineProperty.value @ bootstrap 36e7a445048220614319:724
(anonymous) @ bootstrap 36e7a445048220614319:724
</details>

## Considerations

For whatever reason **write** operations work fine with these rules. I tested this by `enablePersistence` and then using the cached data to invoke write operations. Of course I couldn't tell those operations worked without refreshing the page to reload the cache or via checking my [firestore DB console](https://console.firebase.google.com).

Point being that **write** operations (update, delete, create) got all the way up to firestore in the cloud and back down to the client via the cache. I presume it's from cache since any **read** attempt fails.

The only way I could get **read** to work was to specifically break out the rules so that `resource.data.roles[request.auth.uid]` is **only** invoked at the `stories/{story}` scope and `get(/databases/$(database)/documents/stories/$(story)).data.roles[request.auth.uid]` is **only** invoked at the `comments/{comment}` scope even taking into account the [SO issue][so-issue].

This scope requirement is completely mind boggling.

[so-issue]: https://stackoverflow.com/questions/46629170/firestore-security-rule-get-not-work

**Comment by samtstern** [comment]
@ralphsmith80 before I digest this I just want to say thank you for taking the time to write such a detailed report!  I will take a look.

**Comment by samtstern** [comment]
Looks like part of the problem is my change here:
https://github.com/firebase/snippets-rules/commit/35476a30dc022f29d9c1d3df8cbd44c35a3f0f48#diff-5d2bf3367719655275c45e9735d0bb14

When you're making a new story then the simplification to `resource.data` saves the `get()` request but as you noted this does not work for comments.

**Comment by samtstern** [comment]
@ralphsmith80 as a start, what do you think about this fix to undo the major bug here:
https://github.com/firebase/snippets-rules/pull/5/files

**Comment by ralphsmith80** [comment]
@samedson looks good to me. I validated it in my environment. Your version makes it a little more clear that getting the `resource` needs to be distinct for each case. 

In the documentation or a comment in the snippet it's probably worth noting `get(/databases/$(database)/documents/stories/$(story)).data.roles[request.auth.uid]` doesn't work at the story level. Based on the stackoverflow issue I referenced I expect there's a technical issue with `get` that's still being worked out.

A comment to that affect could save others wasted time if they see the simplification you made.
